pkg/services/screenshot/singleflight.go

```diff
@@ -0,0 +1,47 @@
+package screenshot
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/sync/singleflight"
+)
+
+type captureFunc func(ctx context.Context, opts ScreenshotOptions) (*Screenshot, error)
+
+// SingleFlight is an interface for single flighters. It prevents multiple screenshots of the same scene
+// from being taken at the same time, instead taking the screenshot once and sharing it with all callers
+// that requested the same scene.
+//
+//go:generate mockgen -destination=singleflight_mock.go -package=screenshot github.com/grafana/grafana/pkg/services/screenshot SingleFlight
+type SingleFlight interface {
+	// Do returns a screenshot for the options or an error. It prevents multiple screenshots with
+	// the same options from being taken at the same time. Instead, the screenshot is taken once
+	// and shared with all callers that requested a screenshot with the same options.
+	Do(ctx context.Context, opts ScreenshotOptions, fn captureFunc) (*Screenshot, error)
+}
+
+type SingleFlighter struct {
+	f singleflight.Group
+}
+
+func NewSingleFlight() SingleFlight {
```

Comment:
Do you need this constructor? The concrete type is already exported.

Reply:
It's not needed for initialization, but I thought it would be nice when constructing to have:

```
s = screenshot.NewManagedScreenshotService(
    screenshot.NewHeadlessCaptureService(ds, rs),
    screenshot.NewSingleFlight(),
    screenshot.NewFixedTokenBucket(cfg.UnifiedAlerting.Screenshots.MaxConcurrentScreenshots),
    screenshot.NewInmemCacheService(screenshotCacheTTL, r),
    r)
```

instead of something like this:

```
s = screenshot.NewManagedScreenshotService(
    screenshot.NewHeadlessCaptureService(ds, rs),
    &screenshot.SingleFlight{},
    screenshot.NewFixedTokenBucket(cfg.UnifiedAlerting.Screenshots.MaxConcurrentScreenshots),
    screenshot.NewInmemCacheService(screenshotCacheTTL, r),
    r)
```